File: web/src/schemas/task.ts
// Tasks, their runs, and the durable HITL inquiries a run can block on.
import { z } from 'zod'

// assistant/tasks/model.py RunStatus.
export const RunStatus = z.enum(['running', 'needs_input', 'completed', 'failed', 'cancelled'])
export type RunStatus = z.infer<typeof RunStatus>

// model.py pins this to schedule | once | manual, but run records are persisted
// across versions and no view reads the field, so an older value must not throw.
export const RunTrigger = z.string()
export type RunTrigger = z.infer<typeof RunTrigger>

export const Run = z.object({
  id: z.string(),
  task_id: z.string(),
  status: RunStatus,
  trigger: RunTrigger,
  started_at: z.string().nullable(),
  ended_at: z.string().nullable(),
  summary: z.string().nullable(),
  error: z.string(),
  seen: z.boolean(),
})
export type Run = z.infer<typeof Run>

// assistant/tasks/model.py ScheduleKind, normalised to {kind, at, cron}.
// Never null: model.py:87 defaults it and normalize_schedule(None) yields manual.
//
// All three fields are declared because ScheduleField.svelte reads `at` and
// `cron`, and the gateway's model is the contract — a field it does not declare
// never reaches the client. Both are `.nullish()`: store.py create_task accepts a
// schedule dict unnormalised (`schedule or manual_schedule()`), so a persisted row
// can be missing a key normalize_schedule would have filled. Still a looseObject,
// so a shape the backend later grows is rendered rather than rejected.
export const Schedule = z.looseObject({
  kind: z.enum(['manual', 'once', 'cron']),
  at: z.string().nullish(),
  cron: z.string().nullish(),
})
export type Schedule = z.infer<typeof Schedule>

export const Task = z.object({
  id: z.string(),
  name: z.string(),
  prompt: z.string(),
  model: z.string().nullable(),
  description: z.string(),
  schedule: Schedule,
  schedule_desc: z.string(),
  paused: z.boolean(),
  starred: z.boolean(),
  next_run_at: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
  last_run: Run.nullable(),
  unread: z.number(),
  needs_input: z.boolean(),
})
export type Task = z.infer<typeof Task>

// GET /tasks/{id} is the same row plus its run history.
export const TaskWithRuns = Task.extend({ runs: z.array(Run) })
export type TaskWithRuns = z.infer<typeof TaskWithRuns>

export const TaskList = z.object({ tasks: z.array(Task) })
export type TaskList = z.infer<typeof TaskList>

export const TaskEnvelope = z.object({ task: TaskWithRuns })
export type TaskEnvelope = z.infer<typeof TaskEnvelope>

// Both run routes answer through tasks_service.get_run, which stamps the owning
// task's name onto the run view (tasks_service.py:354) — the run header reads it.
export const RunDetail = Run.extend({ task_name: z.string() })
export type RunDetail = z.infer<typeof RunDetail>

export const RunDetailEnvelope = z.object({ run: RunDetail })
export type RunDetailEnvelope = z.infer<typeof RunDetailEnvelope>

export const RunList = z.object({ runs: z.array(Run) })
export type RunList = z.infer<typeof RunList>

export const Inquiry = z.object({
  id: z.string(),
  task_id: z.string(),
  chat: z.string(),
  kind: z.string(),
  text: z.string(),
  detail: z.string(),
  options: z.array(z.string()),
  created_at: z.string(),
  root_id: z.string().nullable(),
  task_title: z.string(),
  run_id: z.string().nullable(),
})
export type Inquiry = z.infer<typeof Inquiry>

export const InquiryList = z.object({ pending: z.array(Inquiry) })
export type InquiryList = z.infer<typeof InquiryList>

// hitl/desktop.py pending_list — this profile's own registry. `detail` and
// `options` are nullable because hitl/base.py Question defaults both to None (no
// secondary context; no options means a free-text answer). They were declared
// plain strings here, which made every prompt without a detail fail safeParse —
// and the strip catches that and renders an empty list, so transient permission
// prompts never appeared. The OpenAPI gate now holds both sides to this.
export const HitlQuestion = z.object({
  id: z.string(),
  text: z.string(),
  detail: z.string().nullable(),
  options: z.array(z.string()).nullable(),
  kind: z.string(),
  path: z.string(),
})
export type HitlQuestion = z.infer<typeof HitlQuestion>

export const HitlPending = z.object({ pending: z.array(HitlQuestion) })
export type HitlPending = z.infer<typeof HitlPending>

// POST /tasks answers with a freshly built row — tasks_service.create_task
// returns _task_row() without the `runs` key GET /tasks/{id} adds.
export const NewTaskEnvelope = z.object({ task: Task })
export type NewTaskEnvelope = z.infer<typeof NewTaskEnvelope>

File: src/assistant/gateway/schemas/task.py
"""Tasks, their runs, and the durable HITL inquiries a run can block on.

Mirrors web/src/schemas/task.ts. Field lists come from that file, which was
validated against real responses; the bodies themselves are built by
gateway/tasks_service.py (_run_view, _task_row).
"""

from typing import Literal

from pydantic import BaseModel

RunStatusOut = Literal["running", "needs_input", "completed", "failed", "cancelled"]


class RunOut(BaseModel):
    """One run, as tasks_service._run_view builds it.

    ``trigger`` is a bare str, not the Literal tasks/model.py pins it to: run
    records outlive the version that wrote them, and no view reads the field, so
    an older value must be rendered rather than rejected.
    """

    id: str
    task_id: str
    status: RunStatusOut
    trigger: str
    started_at: str | None
    ended_at: str | None
    summary: str | None
    error: str
    seen: bool


class ScheduleOut(BaseModel):
    """A task's schedule, canonicalised to {kind, at, cron} by
    tasks/model.py normalize_schedule.

    ``at`` and ``cron`` carry defaults rather than being required-nullable
    because tasks/store.py create_task accepts a schedule dict unnormalised
    (``schedule or manual_schedule()``), so a persisted row can be missing a key
    that normalize_schedule would have filled. Both are declared: the schedule
    editor reads them (components/task/ScheduleField.svelte), and a model that
    did not declare them would silently drop them from the response.
    """

    kind: Literal["manual", "once", "cron"]
    at: str | None = None
    cron: str | None = None


class TaskOut(BaseModel):
    """One task row for the drawer — tasks_service._task_row."""

    id: str
    name: str
    prompt: str
    model: str | None
    description: str
    schedule: ScheduleOut
    schedule_desc: str
    paused: bool
    starred: bool
    next_run_at: str | None
    created_at: str
    updated_at: str
    last_run: RunOut | None
    unread: int
    needs_input: bool


class TaskWithRunsOut(TaskOut):
    """The same row plus its run history, as get_task returns it."""

    runs: list[RunOut]


class TaskListResponse(BaseModel):
    """GET /api/p/{pid}/tasks — needs-input first, then newest."""

    tasks: list[TaskOut]


class TaskEnvelopeResponse(BaseModel):
    """GET and PATCH /api/p/{pid}/tasks/{task_id} — both answer through
    get_task, so both carry the run history."""

    task: TaskWithRunsOut


class NewTaskEnvelopeResponse(BaseModel):
    """POST /api/p/{pid}/tasks — create_task answers with a freshly built row,
    which has no ``runs`` key yet."""

    task: TaskOut


class RunDetailOut(RunOut):
    """tasks_service.get_run stamps the owning task's name onto the run view —
    the run header reads it."""

    task_name: str


class RunDetailEnvelopeResponse(BaseModel):
    """GET /api/p/{pid}/runs/{run_id} and POST /api/p/{pid}/tasks/{id}/run."""

    run: RunDetailOut


class RunListResponse(BaseModel):
    """GET /api/p/{pid}/tasks/{task_id}/runs — the history without task_name,
    since the caller already knows the task."""

    runs: list[RunOut]


class InquiryOut(BaseModel):
    """One open HITL inquiry. The last three fields resolve the asking run back
    to its task so the strip can label and link the source."""

    id: str
    task_id: str
    chat: str
    kind: str
    text: str
    detail: str
    options: list[str]
    created_at: str
    root_id: str | None
    task_title: str
    run_id: str | None


class InquiryListResponse(BaseModel):
    """GET /api/p/{pid}/inquiries/pending."""

    pending: list[InquiryOut]
